Control.cpp

```diff
@@ -45,61 +50,284 @@ double Control::getAngleError(const double current_angle_x, const double desire)
+  else if (speed > Common::kLimit_sup_pwm) {
+    speed = Common::kLimit_sup_pwm;
   }
 }
 
 double Control::getNewDesireLeft(double new_desire) {
   new_desire = new_desire - kDegrees90;
   if (new_desire < 0) {
-    new_desire = new_desire + kDegrees360;
+    new_desire += kDegrees360;
   }
   return new_desire;
 }
 
 double Control::getNewDesireRight(double new_desire) {
-  new_desire = new_desire + kDegrees90;
+  new_desire += kDegrees90;
   if (new_desire > kDegrees360) {
-    new_desire = new_desire - kDegrees360;
+    new_desire -= kDegrees360;
   }
   return new_desire;
 }
 
 bool Control::detectRamp() {
-  return (bno_.getAngleY() < -kLimitInfDegrees 
-  && bno_.getAngleY() > -kLimitSupDegrees 
-  && bno_.getAngleZ() < kRangeAngleZ 
-  && bno_.getAngleZ() > -kRangeAngleZ);
+  return (bno_->getAngleY() < -kLimitInfDegrees 
+  && bno_->getAngleY() > -kLimitSupDegrees 
+  && bno_->getAngleZ() < kRangeAngleZ 
+  && bno_->getAngleZ() > -kRangeAngleZ);
 }
 
 bool Control::bumperLevel1() {
-  return (bno_.getAngleY() < -kLimitInfBumper1 
-  && bno_.getAngleY() > -kLimitSupBumper1 
-  && bno_.getAngleZ() > kRangeAngleZ 
-  || bno_.getAngleZ() < -kRangeAngleZ);
+  if (bno_->getAngleY() < -kLimitInfBumper1 && bno_->getAngleY() > -kLimitSupBumper1) {
+    if (bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 1.
+    }
+    else {
+      return false; // It's a ramp.
+    }
+  }
+  else {
+    return false; // It's another bumper level.
+  }
 }
 
 bool Control::bumperLevel2() {
-  return (bno_.getAngleY() < -kLimitSupBumper1 
-  && bno_.getAngleY() > -kLimitSupBumper2 
-  && bno_.getAngleZ() > kRangeAngleZ 
-  || bno_.getAngleZ() < -kRangeAngleZ);
+  if(bno_->getAngleY() < -kLimitSupBumper1 && bno_->getAngleY() > -kLimitSupBumper2) {
+    if(bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 2.
+    }
+    else {
+    return false; // It's a ramp.
+    }
+  }
+  else {
+  return false; // It's another bumper level.
+  }
 }
 
 bool Control::bumperLevel3() {
-  return (bno_.getAngleY() < -kLimitSupBumper2 
-  && bno_.getAngleY() > -kLimitSupBumper3 
-  && bno_.getAngleZ() > kRangeAngleZ 
-  || bno_.getAngleZ() < -kRangeAngleZ);
+  if(bno_->getAngleY() < -kLimitSupBumper2 && bno_->getAngleY() > -kLimitSupBumper3) {
+    if (bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 3.
+    }
+    return false; //It's a ramp.
+  } 
+  return false; // It's another bumper level.
+}
+
+double Control::getPwmBNOLeft(const double desire) { // Positive.
+    double pwm_left = 0;
+    double pwm_right = 0;
+    double pwm_left_final = 0;
+    double pwm_right_final = 0;
+    double pwm_right_BNO = 0;
+    double pwm_left_BNO = 0;
+    double current_angle_x = bno_->getAngleX(); 
+    const double errorBNO = getAngleError(current_angle_x, desire);
```

Comment:
You can also do: 

const double errorBNO = (getAngleError(bno_->getAngleX(), desire);